### PR TITLE
Fix AvgTests error on -0.0 avg

### LINKED ISSUE #113225: [CI] AvgTests testFold {TestCase=<double> #2} failing
**Build Scans:**
- [elasticsearch-periodic-platform-support #4091 / debian-10_platform-support-unix](https://gradle-enterprise.elastic.co/s/kxw7akip5ph2m)
- [elasticsearch-pull-request #32022 / part-3](https://gradle-enterprise.elastic.co/s/d6zixmta5z7ye)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:test" --tests "org.elasticsearch.xpack.esql.expression.function.aggregate.AvgTests.testFold {TestCase=<double> #2}" -Dtests.seed=F77328418405372D -Dtests.locale=sd-Deva -Dtests.timezone=America/Eirunepe -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testFold%20%7BTestCase%3D%3Cdouble%3E%20%232%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.expression.function.aggregate.AvgTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <0.0>
     but: was <-0.0>
```

**Issue Reasons:**
- [main] 2 consecutive failures in test testFold {TestCase=<double> #2}
- [main] 2 failures in test testFold {TestCase=<double> #2} (100.0% fail rate in 2 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### LINKED ISSUE #114175: [CI] AvgTests testFold {TestCase=<double> #7} failing
**Build Scans:**
- [elasticsearch-periodic-platform-support #4272 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/zvqe5znyeafv2)
- [elasticsearch-intake #10355 / part3](https://gradle-enterprise.elastic.co/s/l3qmuokwlav6s)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:test" --tests "org.elasticsearch.xpack.esql.expression.function.aggregate.AvgTests.testFold {TestCase=<double> #7}" -Dtests.seed=5F18D190991DD34C -Dtests.locale=ks-Arab -Dtests.timezone=Asia/Dhaka -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testFold%20%7BTestCase%3D%3Cdouble%3E%20%237%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.expression.function.aggregate.AvgTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <0.0>
     but: was <-0.0>
```

**Issue Reasons:**
- [main] 2 failures in test testFold {TestCase=<double> #7} (50.0% fail rate in 4 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### PR DESCRIPTION
**Pull request by ivancea**
Fixes https://github.com/elastic/elasticsearch/issues/113225
Fixes https://github.com/elastic/elasticsearch/issues/114175

### CODE REVIEW

**Comment by not-napoleon** [comment]
Maybe I'm reading this wrong, but it looks to me like this will accept `-0.0` as the correct output.  We really shouldn't be returning negative zeros from any operations.  (In fact, probably `AbstractAggregatorTestCase#extractResultsFromAggregator` should assert that we never return a forbidden double value).

**Comment by ivancea** [comment]
> Maybe I'm reading this wrong, but it looks to me like this will accept `-0.0` as the correct output. We really shouldn't be returning negative zeros from any operations.

@not-napoleon `TestCaseSupplier` provides -0.0 values, so if it can be an input, it should also be an output.

In ES, I see this: https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html#_which_type_should_i_use (Just above that anchor)

> The double, float and half_float types consider that -0.0 and +0.0 are different values.

So if ES also allows it, I would allow it too. And for this case, the avg of -0.0 should be -0.0 too (IMO) to be as precise as possible.

I don't have an specific opinion on this topic, I'm just basing my thought process on what we currently have

**Review by not-napoleon — CHANGES_REQUESTED** [review]
> to be as precise as possible.

It's not more precise, it's just more confusing. The IEEE spec for floating points says that the two zeros should be considered equal when possible.  Java honors this with it's binary comparison operators (`==` and friends), but not with `Double.equals()`.  You can read the docs on the [double class](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html) to get an understanding for why this is the case.  Practically, this means that if we start allowing negative zeros in the language, we need to be VERY clear on where they compare via `==` and where they compare via `Double.equals()` or `Number.equals()`.  It becomes extremely trappy for users if that behavior is not entirely consistent.

We made the decision early on to not allow `-0.0`, positive or negative infinity, or `NaN` as possible values in ESQL.  We should be consistent with that here.  /cc @costin

**Comment by not-napoleon** [comment]
> so if it can be an input, it should also be an output.

It's an input on the test precisely to test that it isn't an output.

**Comment by nik9000** [comment]
> We made the decision early on to not allow `-0.0`, positive or negative infinity, or `NaN` as possible values in ESQL. We should be consistent with that here. /cc @costin

I think we should be consistent with this decision. Our functions should never receive any of those forbidden doubles. We could (should?) add an assertion to the generated code that reads `double`s for EVAL and STATS that `assert`s that we don't receive those numbers. They should always come in as `null`. And they should always be emitted as `null`.

ES might support these but they are all `null` to ESQL. One day we may change this but that's a difficult thing to change at the moment.

**Comment by not-napoleon** [comment]
> ES might support these but they are all null to ESQL. One day we may change this but that's a difficult thing to change at the moment.

Well, the infinities and `NaN` are null; `-0.0` is `0`.

**Comment by ivancea** [comment]
Ok, so:
- To ensure it's not an output, I'll add a check to tests like we have to NaN and infinities
- I'll revert the changes on AvgTests for now

Then, if we have to keep the -0.0 as a potential input in tests, but forbid it as an output, as many functions are currently returning it (Max, MvMax, Min, Values...), I'll fix them by doing a conversion (-0.0 -> 0.0).
I could add it in the agg/evaluator implementers, but it would affect even functions never emitting -0.0. So I'll see first how many functions are failing with this

**Comment by ivancea** [comment]
@not-napoleon After some checks and talks about this:
- Added some "discoveries" about -0.0 here, to tackle whenever we want to work on this: https://github.com/elastic/elasticsearch/issues/104866#issuecomment-2407370445
- Restored this PR with the Avg fix. It does nothing around -0.0 directly, so I think it's safe to merge whatever the decision we take. Right now, it's agnostic.
  - I left the little refactor on tests: removing repeated code, and adding test cases of single values to aggs, as they are handled differently (This test was failing 1/1000 times because of this)

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java [hunk 1]
```diff
@@ -750,6 +752,26 @@ public static void testFunctionInfo() {
         assertEquals(returnFromSignature, returnTypes);
     }
 
+    @SuppressWarnings("unchecked")
+    protected final void assertTestCaseResultAndWarnings(Object result) {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:756]: Javadoc would be nice here.

**Review by not-napoleon — APPROVED** [review]
This is fine for now.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java (hunk 1 above)
Comment (ivancea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:762]: Added!

**Pull request merged**